### PR TITLE
Implementation of lofreq callparallel to sarek as variant callers

### PR DESCRIPTION
**Pull request by AitorPeseta**
#1588 

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

We wanted to ask you for a code review in this PR. We are updating Sarek with new modules: lofreq somatic (paired sample variant calling) and lofreq callparallel (tumor-only variant calling). Therefore, there will be two new subworkflows: one for somatic and the other one for tumor-only.

In order to these modules fitting in Sarek, we have made some changes in them in this fork. Consequently, they are not exactly the same modules from nf-core modules repo (they will be updated when we finish this). For this reason, we will be working in parallel to update these modules in nf-core/modules.

To sum up, we wanted to ask for a code review, specifically in these new lofreq subworkflows.

### CODE REVIEW

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

File: README.md [hunk 1]
```diff
@@ -42,7 +42,7 @@ It's listed on [Elixir - Tools and Data Services Registry](https://bio.tools/nf-
 Depending on the options and samples provided, the pipeline can currently perform the following:
 
 - Form consensus reads from UMI sequences (`fgbio`)
-- Sequencing quality control and trimming (enabled by `--trim_fastq`) (`FastQC`, `fastp`)
+- Sequencing quality control and trimming (enabled by `--trim_fastq`) (`FastQC`, `fastp`,`bedtools`)
```
Comment (FriederikeHanssen) [inline README.md:45]: bedtools is not used for the read data but preparing the intervals file, right?

File: docs/output.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
     - [Trim adapters](#trim-adapters)
     - [Split FastQ files](#split-fastq-files)
     - [UMI consensus](#umi-consensus)
+    - [Bedtools](#bedtools)
```
Comment (FriederikeHanssen) [inline docs/output.md:19]: the bedtools output is changing the intervals file + there is only one per run/it's the same for all samples. IMO this should go to the reference directory, not the sample directories.

File: docs/output.md [hunk 3]
```diff
@@ -158,6 +160,22 @@ These files are intermediate and by default not placed in the output-folder kept
...
+[Bedtools](https://github.com/arq5x/bedtools2) utilities are a swiss-army knife of tools for a wide-range of genomics analysis tasks. The most widely-used tools enable genome arithmetic. Bedtools allows one to intersect, merge, count, complement, and shuffle genomic intervals from multiple files in widely-used genomic file formats such as BAM, BED, GFF/GTF, VCF.
+While each individual tool is designed to do a relatively simple task (e.g., intersect two interval files), quite sophisticated analyses can be conducted by combining multiple bedtools operations on the UNIX command line.
+
+<details markdown="1">
+<summary>Output files for all samples</summary>
+
+**Output directory: `{outdir}/reports/bedtools/`**
+
+- `<sample>.bed`
+  - New .bed file with the news changes.
```
Comment (FriederikeHanssen) [inline docs/output.md:174]:
```suggestion
  - New .bed file with the new changes.
```

can you describe the actual changes?

File: nextflow_schema.json [hunk 4]
```diff
@@ -111,15 +111,15 @@
...
+                    "pattern": "^((ascat|bcfann|cnvkit|controlfreec|deepvariant|freebayes|haplotypecaller|sentieon_dnascope|sentieon_haplotyper|manta|merge|mpileup|msisensorpro|mutect2|lofreq|ngscheckmate|sentieon_dedup|snpeff|strelka|tiddit|vep)?,?)*(?<!,)$"
```
Comment (FriederikeHanssen) [inline nextflow_schema.json:115]:
```suggestion
                    "pattern": "^((ascat|bcfann|cnvkit|controlfreec|deepvariant|freebayes|haplotypecaller|lofreq|sentieon_dnascope|sentieon_haplotyper|manta|merge|mpileup|msisensorpro|mutect2|ngscheckmate|sentieon_dedup|snpeff|strelka|tiddit|vep)?,?)*(?<!,)$"
```

File: subworkflows/local/bam_variant_calling_tumor_only_lofreq/main.nf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    // Combine cram and intervals for spread and gather strategy
+    input_intervals = input.combine(intervals_final)
+        // Move num_intervals to meta map
+        .map {meta, tumor_cram, tumor_crai, intervals, num_intervals -> [meta + [ num_intervals:num_intervals ], tumor_cram, tumor_crai, intervals]}
+
+    LOFREQ(input_intervals, fasta, fai) // Call variants with LoFreq
+
+    vcf = Channel.empty().mix(LOFREQ.out.vcf)
+        .map{ meta, vcf -> [ meta + [ variantcaller:'lofreq' ], vcf ] }
+    versions = versions.mix(LOFREQ.out.versions)
```
Comment (FriederikeHanssen) [inline subworkflows/local/bam_variant_calling_tumor_only_lofreq/main.nf:36]: the versions for SORT and MERGE intervals also need to be collected.

**Review by FriederikeHanssen — CHANGES_REQUESTED** [review]
- the docs/images/mqc* files shouldn't be changed by this PR. Can you remove them here?
- something fishy is going on with the main.nf. If you didn't change it, it shouldn't be in this commit
- modules/cnvkit* should not have been changed by this PR
- General question: I really like the addition of bedtools.  You put the `bedtools` modules as run by default ( they are part of the list of `skip_tools` ) . Yet in your current implementation they are only used for LoFreq. 
I would argue that sorting and merging the bedfiles is of general usage to many tools. there are quite a few variant callers in this pipeline that will fail unless the bed file is sorted and merged ( it might even be worth a separate PR to keep things small and contained). Anotehr feature that may be useful of you implement bedtools is padding of each side by a number of basepairs as recommended by GATK guideline for preprocessing and unpadding before doing the variant calling.

**Review by FriederikeHanssen — CHANGES_REQUESTED** [review]
Can you also please add a test that tests this subworkflow?

File: README.md (hunk 1 above)
Comment (AitorPeseta) [reply README.md:45]: You're right, on second thought, I think the Process BAM file option is better.

File: docs/output.md (hunk 2 above)
Comment (AitorPeseta) [reply docs/output.md:19]: Then, would the best option be Map to Reference?

File: docs/output.md (hunk 3 above)
Comment (AitorPeseta) [reply docs/output.md:174]: Done :)

File: subworkflows/local/bam_variant_calling_tumor_only_lofreq/main.nf (hunk 5 above)
Comment (AitorPeseta) [reply subworkflows/local/bam_variant_calling_tumor_only_lofreq/main.nf:36]: Done :)

**Comment by AitorPeseta** [comment]
Good morning @FriederikeHanssen, I have a question. I’m working on the test yml file, but since it's my first time, I don't know where to get the md5sum number from. Thank you very much.

**Comment by FriederikeHanssen** [comment]
with pytest you sadly need to compute them with good old `md5sum` 😭 nf-test is on the horizon somewhere which will make this easier in the future

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

**Comment by nevinwu** [comment]
Finally, we have decided to **remove the processing of the interval file** with `bedtools` in the **subworkflow**. We agree that it would make more sense to add this processing in a "general" step for all the callers before doing the variant calling. So, we'll talk about working on this in the future in a separate PR.

Furthermore, removing this step in the subworkflow allows the usage of `lofreq` without using an intervals file.

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

**Comment by nevinwu** [comment]
We have added test for the `lofreq` subworkflow. However we would like to ask for some help in 2 points:

1. In the first test (`lofreq` with intervals), we don't know why, **in the `results/csv/variantcalled.csv` file**, the **sample appears twice in two sepparated lines**. Maybe it's something related with the use of intervals and `CHANNEL_VARIANT_CALLING_CREATE_CSV`? Any idea of what might be happening?
2. Since `VCFTOOLS_TSTV_COUNT` fails with `lofreq`, we had to add `errorStrategy = 'ignore'` in the `lofreq` config file with this module (`conf/modules/lofreq.config`). Don't know if is there any other way to deal with this.

Thanks again.

**Comment by AitorPeseta** [comment]
@nf-core-bot fix linting

**Comment by nevinwu** [comment]
We have added `tests` and implemented the use of `intervals` with `lofreq callparallel` (which was missing previously). Any further feedback would be apreciated.

File: docs/output.md [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# nf-core/sarek: Output <!-- omit in toc -->
+º1# nf-core/sarek: Output <!-- omit in toc -->
```
Comment (maxulysse) [inline docs/output.md:1]:
```suggestion
# nf-core/sarek: Output <!-- omit in toc -->
```

File: docs/output.md [hunk 7]
```diff
@@ -195,6 +196,8 @@ The alignment files (BAM or CRAM) produced by the chosen aligner are not publish
   - BAM file and index
   </details>
 
+</details>
+
```
Comment (maxulysse) [inline docs/output.md:200]:
```suggestion
```

**Comment by maxulysse** [comment]
Why have the FASTQC png files changed?

Can you update CHANGELOG, I think it's almost good to go

**Comment by nevinwu** [comment]
> Why have the FASTQC png files changed?

To be honest, I don't really know why the FASTQC `png` files have changed. I think my collegue Àitor unintentionally touched them, git was saying they were changed, so he had to download the ones in `master` and replace the files in his fork. I don't know if just this change can be reverted or undonde easily.

> Can you update CHANGELOG, I think it's almost good to go

CHANGELOG updated, please let me know if there is anything left to be done. :)

**Comment by nevinwu** [comment]
@nf-core-bot fix linting

**Review by FriederikeHanssen — DISMISSED** [review]
Looks really great!!!

One question: What is going on with the main.nf and genemetrics/* files? Why are they changed? The main.nf should at most have punctual inclusion. 

One feature request: can you please update the metromap and pipeline overview image?

Other than that ready to merge form my side

**Comment by nevinwu** [comment]
Files that shouln't have been changed should be fixed by now. 

I have updated the `sarek_workflow` image (I have used `inkscape` if that's okay), but I don't think I have the skills to update the `sarek_subway` one. Could I get a little help with that please? I am a total newbie to all of this image-editing thing.

**Comment by maxulysse** [comment]
Can you create an issue, and I'll update the metro map at a later date (before the release).
I noticed some weird changes with the new version of the workflow, but don't worry about it, I'll have a look at the same time.

**Review by maxulysse — APPROVED** [review]
